### PR TITLE
reduce network usages

### PR DESCRIPTION
**Pull request by ugatenio**
# Description

See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-dashboard/src/main/js/components/karaoke/components/Pipeline.jsx [hunk 1]
```diff
@@ -114,13 +126,7 @@ export default class Pipeline extends Component {
                 logger.debug('Need to set new Run. Happens when e.g. re-run.');
                 nextProps.augmenter.setRun(nextProps.run);
             }
-            debounce(() => {
-                if (KaraokeConfig.getPreference('runDetails.pipeline.karaoke').value !== 'never') {
-                    logger.debug('re-setting karaoke mode.');
-                    this.karaoke = true;
-                }
-                this.pager.fetchNodes({ node: nextProps.params.node });
-            }, 200)();
```
Comment (ugatenio) [inline blueocean-dashboard/src/main/js/components/karaoke/components/Pipeline.jsx:123]: @bitwiseman same problem as described in: https://github.com/jenkinsci/blueocean-plugin/pull/2104
there is other misuse of debounce in this file
but this one specific cause the flooding of GET request

**Comment by ugatenio** [comment]
this is just part of all the same requests
Screen Shot 2020-09-30 at 13 51 47

**Comment by eplodn** [comment]
Probably related to JENKINS-55849.

Every load of `/nodes?limit=10000` seems to cause a scroll to the end. And there are too many loads of that url without any limit. I can actually see 5-8 outstanding requests to that URL _at the same time_.

**Comment by eplodn** [comment]
image
Here about 50 requests issued at the same 5-second interval.

**Comment by eplodn** [comment]
There's also something that looks like a streaming subscription to Jenkins events:

```event: open
data: {"dispatcherId":"jenkins-blueocean-core-js-1601567338925-k1ieqj","dispatcherInst":1837042023}

event: pipeline
data: {"jenkins_event_uuid":"dbd1144d-43ac-42cf-96cb-5f679fb05a65","pipeline_job_name":"company/project/PR-24813","pipeline_run_id":"11","jenkins_event_timestamp":"1601567709426","pipeline_step_name":"sh","pipeline_step_stage_id":"2259","pipeline_step_is_paused":"false","pipeline_step_flownode_id":"2259","jenkins_event":"pipeline_stage","pipeline_step_stage_name":"Store kernel log...","pipeline_context":"2/47/48/49/50/51/52/53/54/55/56/888/889/890/891/959/1085/1211/1212/1305/1427/1450/1451","jenkins_instance_url":"https://jenkins.dev.company.net/","jenkins_channel":"pipeline"}

event: pipeline
data: {"jenkins_event_uuid":"538d7f7d-a1d9-4d53-925d-211cfadd6209","pipeline_job_name":"company/project/PR-24813","pipeline_run_id":"11","jenkins_event_timestamp":"1601567709734","pipeline_step_name":"sh","pipeline_step_stage_id":"2260","pipeline_step_is_paused":"false","pipeline_step_flownode_id":"2260","jenkins_event":"pipeline_stage","pipeline_step_stage_name":"Storing journalctl system logs...","pipeline_context":"2/47/48/49/50/51/52/53/54/55/56/888/889/890/891/959/1085/1211/1212/1305/1427/1450/1451","jenkins_instance_url":"https://jenkins.dev.company.net/","jenkins_channel":"pipeline"}

event: pipeline
data: {"jenkins_event_uuid":"042da946-3f9e-4dcd-816e-df7fbc2f2792","pipeline_job_name":"company/project/PR-24813","pipeline_run_id":"11","jenkins_event_timestamp":"1601567710601","pipeline_step_name":"sh","pipeline_step_stage_id":"2261","pipeline_step_is_paused":"false","pipeline_step_flownode_id":"2261","jenkins_event":"pipeline_stage","pipeline_step_stage_name":"Store remote kernel log...","pipeline_context":"2/47/48/49/50/51/52/53/54/55/56/888/889/890/891/959/1085/1211/1212/1305/1427/1450/1451","jenkins_instance_url":"https://jenkins.dev.company.net/","jenkins_channel":"pipeline"}

event: pipeline
data: {"jenkins_event_uuid":"a7151062-c7d2-4d91-aa51-57a7cb4a1ebb","pipeline_job_name":"company/project/PR-25110","pipeline_run_id":"4","jenkins_event_timestamp":"1601567710716","pipeline_step_name":"retry","pipeline_step_stage_id":"97","pipeline_step_flownode_id":"102","jenkins_event":"pipeline_block_end","pipeline_step_stage_name":"changing ownership","pipeline_context":"2/47/48/49/50/51/52/53/54/55/56/57/58/59/60/98/99","jenkins_instance_url":"https://jenkins.dev.company.net/","jenkins_channel":"pipeline"}
```
Once the event is received, blueocean issues a request on build status — of _completely unrelated build_ on a _completely unrelated branch/PR_. This causes the blueocean interface to never finish loading my page.  

We have tens of builds going at the same time, and hundreds of blueocean pages opened across the company. I don't know why it is a good idea to have these hundreds of pages refresh all the time.

**Comment by ugatenio** [comment]
also related to this code:
https://github.com/jenkinsci/blueocean-plugin/blob/77960a8dc21840ae5f6df784a26f91e09d7da621/blueocean-dashboard/src/main/js/components/karaoke/rest/KaraokeApi.js#L92
which seems more complicated to fix.
@bitwiseman please advise

**Comment by cliffmeyers** [comment]
I'm not an active maintainer / contributor to this plugin anymore but here's my 2 cents:
- On the surface this change also looks correct to me.
- If @ugatenio or @eplodn have a way to show before / after of the network traffic related to this change, I think it'd go a lot way to increasing confidence that the change is correct.
- I might suggest fixing each instance of `debounce` misuse in a separate PR just so it's easier to revert / test in the future in case any of the changes create regressions.

**Comment by eplodn** [comment]
@cliffmeyers  we are deploying the changes on a test Jenkins instance  as we would rather not break the production server; the test server has no real slaves and no jobs running in the background so it does not expose the problematic behavior to begin with. Regardless, we'll try to provide the data  as you requested.

File: blueocean-dashboard/src/main/js/components/karaoke/components/Pipeline.jsx [hunk 2]
```diff
@@ -46,6 +46,15 @@ export default class Pipeline extends Component {
...
         this.state = { tailLogs: false };
+
+        this.debounceFetchNodes = debounce((karaokeOut) => {
+            logger.debug('sse fetch it', this.karaoke);
+            if (karaokeOut) {
+                this.pager.fetchNodesOnly({});
+            } else {
+                this.pager.fetchNodes({});
+            }
+        }, 200);
```
Comment (ugatenio) [inline blueocean-dashboard/src/main/js/components/karaoke/components/Pipeline.jsx:57]: @bitwiseman, @ cliffmeyers this made significant improvement in our env.
consider even to increase from 200 to 500-1000

**Pull request merged**